### PR TITLE
Automatic/configurable datasets download. Fixes #376 #409

### LINKED ISSUE #376: Implement automatic download of remote datasets
Whenever a remote algorithm finished execution its outputs become ready for download. The default behavior for Trends.Earth shall be to proceed to download them automatically.

Foreseen workflow is:

- A refresh of the datasets list is initiated (either automatically or by request of the user)
- Datasets that are in a `being generated` state are checked to discover whether they are ready to be downloaded
- trends.earth proceeds to download datasets that are ready, updating their state accordingly

Downloading a dataset shall be made in a background task, in order to prevent locking up the QGIS main UI and allow the user to carry on with other work.

Additionally we shall allow this automatic download behavior to be disabled by the user. This shall be tracked in #373 and is not in scope for this issue.

### PR DESCRIPTION
**Pull request by luipir**
This PR add automatic download of datasets that are in FINISHED or SUCCESS state.
Autodownload can be disabled setting the boolean config parameter:

`trends_earth/advanced/dataset_auto_download`

that is not present is assumed True => auto download
After download download state is updated to the next Dataset refresh from base_data_directory

Downloaded dataset are not downloaded again if already downloaded BTW the original dataset json descriptor is not removed, just ignored is the equivalent downloaded json descriptor is already present.

This PR shoudl fix the following issue/features:
https://github.com/ConservationInternational/trends.earth/issues/376
https://github.com/ConservationInternational/trends.earth/issues/409

### CODE REVIEW

**Comment by luipir** [comment]
rebased including the new @Samweli feature merged in decision-tree branch

File: LDMP/models/datasets.py
Comment (ricardogsilva) [inline LDMP/models/datasets.py:271]: This line is flawed as there are only two arguments to the `format()` function but the string specifies three placeholders. In my case, since I did not assign a name to the sample job I created, it makes the plugin crash

```suggestion
                fileBaseName = u'{}_{}'.format(job.runId, job.scriptName)
```

**Review by ricardogsilva — DISMISSED** [review]
It is great to see our design starting to materialize, nice work!

I have a couple of requests, in addition to the bug which I mention inline:

- If the download is set to be automatic (which is the default behaviour) then the download button should not be shown at all. It should only be shown when the user has disabled automatic downloads (as per #373). Even then, the download button should be disabled until we know that the job has finished running

- When downloading finishes, the download button still remains visible until the next time the remote server is checked. It should be hidden immediately, since we even show a message to the user to let him know the file has been downloaded. Otherwise it becomes confusing for the user, since it seems that the button should be clicked.

- When download finishes, I see two JSON files are present in the filesystem, under `{base-dir}/outputs/SDG15.3.1/{creation-date}`:

  - job file
  - downloaded dataset file

  Seems to me that the job file is no longer relevant, after the dataset has been downloaded. The dataset JSON file seems to contain the relevant details about the dataset and also about the job that originated it. Is this correct? If so, please remove the job file

Additional notes, that we ought to address, in separate issues:

- Since we are not able to get a ganular enough progress report to show the execution percentage when the dataset is being generated (this was confirmed by @azvoleff  in the previous meeting) it would be best to show a message like `PROCESSING` instead of the progress bar. Or maybe show a progress bar with the infinite loading status to denote it is doing something but we can't be more specific about the progress?

- This is a minor nitpick but the widgets move slightly when the download icon is removed. It would be nicer if they didn't.

File: LDMP/models/datasets.py
Comment (Samweli) [inline LDMP/models/datasets.py:95]: I would prefer to define the Origin enum outside the `Dataset` class

File: LDMP/models/datasets.py
Comment (Samweli) [inline LDMP/models/datasets.py:283]: shouldn't we make this a static method?

**Review by Samweli — COMMENTED** [review]
I think we need to choose one naming convention that should be use, I see currently there is a mix of camel case with snake case, also we need to fix typos in documentation as they can mislead the intended interpretation.

**Comment by luipir** [comment]
> * Even then, the download button should be disabled until we know that the job has finished running

no button is available until finished

**Comment by luipir** [comment]
> * When downloading finishes, the download button still remains visible until the next time the remote server is checked. It should be hidden immediately, since we even show a message to the user to let him know the file has been downloaded. Otherwise it becomes confusing for the user, since it seems that the button should be clicked.

just because the datasets item is refreshed at the end of refresh rate... I can disable or hide the download if set auto download

**Comment by luipir** [comment]
> * f so, please remove the job file

ok, I leaved it just in case and traceability

**Comment by luipir** [comment]
> * Or maybe show a progress bar with the infinite loading status

good idea

File: LDMP/models/datasets.py
Comment (luipir) [reply LDMP/models/datasets.py:95]: rationale?

File: LDMP/models/datasets.py
Comment (luipir) [reply LDMP/models/datasets.py:283]: could be static because it's a singleton working on class variable... btw I would avoid to access it without a single instance on which to work. e.g I would leave it as singleton to force user use singleton constructor and explicitly dige the idea they are working on a single instance

**Comment by ricardogsilva** [comment]
> > * Even then, the download button should be disabled until we know that the job has finished running
> 
> no button is available until finished

What I meant was, when implementing the changes for #373 it would be good if the download button then becomes visible nut disabled while the dataset is being prepared (so that the user is able to discover that at some point he will be able to download the dataset manually). Then, when available, the download button becomes enabled, so that the user is able to download it. Finally, when the user has already downloaded, the button is hidden.

But again, this shall be only for when the user has disabled automatic downloads.

**Comment by luipir** [comment]
> becomes visible but disabled while the dataset is being prepared (so that the user is able to discover that at some point he will be able to download the dataset manually)

now it's clear

**Comment by luipir** [comment]
> > > * Even then, the download button should be disabled until we know that the job has finished running
> > 
> > 
> > no button is available until finished
> 
> What I meant was, when implementing the changes for #373 it would be good if the download button then becomes visible but disabled while the dataset is being prepared (so that the user is able to discover that at some point he will be able to download the dataset manually). Then, when available, the download button becomes enabled, so that the user is able to download it. Finally, when the user has already downloaded, the button is hidden.
> 
> But again, this shall be only for when the user has disabled automatic downloads.

should be done now

**Comment by luipir** [comment]
> * This is a minor nitpick but the widgets move slightly when the download icon is removed. It would be nicer if they didn't.

should be done now

**Comment by luipir** [comment]
> * Since we are not able to get a ganular enough progress report to show the execution percentage when the dataset is being generated (this was confirmed by @azvoleff  in the previous meeting) it would be best to show a message like `PROCESSING` instead of the progress bar. Or maybe show a progress bar with the infinite loading status to denote it is doing something but we can't be more specific about the progress?
> *

set "Processing..." and not dynamic progressbar. the tech reason is that any itme in the view is a statically rendered widget that become a working widget (editor) when over it with the mouse. In this way is hard to have widget repaint.

**Comment by luipir** [comment]
> I think we need to choose one naming convention that should be use, I see currently there is a mix of camel case with snake case, also we need to fix typos in documentation as they can mislead the intended interpretation.

I agree... I usually use camel case, but part of the original code is snake case => what to follow?

**Comment by luipir** [comment]
> * f so, please remove the job file

should be done now

**Review by ricardogsilva — CHANGES_REQUESTED** [review]
Some things still need some additional modification:

1. When automatic download is enabled, the download button should never be visible

2. Periodic polling of the remote server does not seem to perform the same actions as manually pressing the refresh button. If I let QGIS automatically poll the remote GEE server then it never becomes aware that the processing is ready and it never downloads the results. I found that I have to explicitly press the 'refresh' button once in order to have QGIS become aware of the availability of the results. Even then, I still need to wait some time before the file is actually downloaded and for the datasets list to be updated.

Here is a small video I made depicting this problem with the automatic polling


https://user-images.githubusercontent.com/732010/117475456-f99e9c80-af53-11eb-87c9-11b4620c25ce.mp4

**Comment by luipir** [comment]
> 2\. Periodic polling of the remote server does not seem to perform the same actions as manually pressing the refresh button. If I let QGIS automatically poll the remote GEE server then it never becomes aware that the processing is ready and it never downloads the results. I found that I have to explicitly press the 'refresh' button once in order to have QGIS become aware of the availability of the results. Even then, I still need to wait some time before the file is actually downloaded and for the datasets list to be updated.

Yes... I realised I forgot to trigger the periodic singleShot....

**Comment by luipir** [comment]
I've rebased after merge of icen PR by @Samweli 
Now fixing the features not yet completed @ricardogsilva

File: LDMP/models/datasets.py
Comment (Samweli) [reply LDMP/models/datasets.py:96]: We should have one place where all the enums are defined, eg in the dataset model class.

Comment (luipir) [reply LDMP/models/datasets.py:96]: I would agree... but the other enum (DatasetStatus) will be removed because unused (as described in the TODO) ... just leaved in case we change idea.
I put the enum in the class just to leave clear the contect of the origin (e.g. is Dataset), otherwise I should name it DatasetOrigin, and IMHO it's more contextual aware having it accessible as class variable of the Dataset class. 
BTW I'm ok also to move outside if this is important for you.
Let mw knoe your decision.... it's up to you.

**Comment by luipir** [comment]
> 1. When automatic download is enabled, the download button should never be visible
> 2. Periodic polling of the remote server does not seem to perform...

done

**Comment by Samweli** [comment]
> 2\. Periodic polling of the remote server does not seem to perform the same actions as manually pressing the refresh button. If I let QGIS automatically poll the remote GEE server then it never becomes aware that the processing is ready and it never downloads the results. I found that I have to explicitly press the 'refresh' button once in order to have QGIS become aware of the availability of the results. Even then, I still need to wait some time before the file is actually downloaded and for the datasets list to be updated.

@luipir I have pulled the latest changes but I still have this issue.

**Comment by luipir** [comment]
> I have pulled the latest changes but I still have this issue.

checking if I missed something

**Review by ricardogsilva — CHANGES_REQUESTED** [review]
I still get some funky behavior with the latest changes:

1. I ask for an algorithm to run
1. Algorithm starts running. I can see that automatic polling of the remote server is running by looking at the QGIS console
1. Sometime later, I receive an email informing me that algorithm execution has finished. Meanwhile QGIS is still not aware of this, which I guess is understandable. I would expect that by the next time Trends.Earth polls the remote server, it also becomes aware that execution is done
1. I can see on the QGIS console that TE does a new poll of the remote server AFTER I have received this email, so I would think the remote server now reports to T.E. that execution is finished. Regardless, T.E. still does not mention that execution is finished
1. After waiting one more minute, T.E does an additional poll of the remote server and now it seems to finally be aware that the remote server has finished execution and proceeds to automatically download the result - this is more than two minutes after I got the email saying that execution is finished
1. I see that automatic download of the result is working now, which is good
1. After having the result downloaded locally, I see that the job file is still present on the base dir. I need to wait another minute, at the time when T.E does another update to get the job file to be deleted. I think the job file should be immediately deleted after the download is done, because we will have the result JSON file now

**Comment by luipir** [comment]
> 3\. I would expect that by the next time Trends.Earth polls the remote server, it also becomes aware that execution is done

me too... but I  update basing on what is the answer from server side (1' is the default polling time). T.E is passive in this case.

**Comment by Samweli** [comment]
> @Samweli are you sure you are using the latest modification? seems it works for @ricardogsilva

@luipir I have pulled the latest changes and it works when I submit new downloads, but the old ones have no progress anymore.

**Comment by luipir** [comment]
> 7\. After having the result downloaded locally, I see that the job file is still present on the base dir. I need to wait another minute, at the time when T.E does another update to get the job file to be deleted. I think the job file should be immediately deleted after the download is done, because we will have the result JSON file now

with latest commit https://github.com/ConservationInternational/trends.earth/pull/412/commits/1e6097ab462ecdadf485297de8033bb077e0d7aa should remove immediatly all json that has been downloaded. Effectively I used the wrong update workflow. just updating the model but not synch datasets with that available in the folder => good catch @ricardogsilva

**Comment by luipir** [comment]
> but the old ones have no progress anymore

@Samweli has been downloaded automatically => no more progress bar? Or I do not understand what you mean? can you share snapshot?

**Comment by luipir** [comment]
> @luipir The download doesn't continue for old datasets and the progress bar is not removed, see ...

first step is to build Datasets form Jobs and Jobs are:
1) or downloaded from server
2) get from cache that is stored in qgis.ini... cache is updated when jobs are downloaded from server.

btw I'll check if there is a cleaning action of jobs from base data directory... I suppose there are old jobs remained when there where no possibility to remove them.
I'll check later... now I've to have a break for bureaucracy matters. In the meantime, please check in base_data_directry/Jobs if there are Jobs that are not listed in cache or available from server

**Comment by luipir** [comment]
@Samweli probably it's time to implement it. https://github.com/luipir/trends.earth/blob/datasets_download_rebased/LDMP/jobs.py#L246
I leaved it becaseu I supposed was not enough important. Case as that you faces is due to the developing situation... shouldn't be a normal case. BTW I'll implement cleaning  of jobs if not present in current list of downloaded jobs from server.
Possible side effect! a Job and related Dataset disappear if not downloaded, btw if not downloaded after 15days probably was not so interesting.

**Comment by Samweli** [comment]
> @Samweli I implemented cleaning of Jobs outside the cache here [luipir@daa8493](https://github.com/luipir/trends.earth/commit/daa84931ea618d80e92f0028749daa165feaf25f)
> I can't cherry-pick it because need to move traverse method in **init** and this is done in the following PR related to the delete of datasets.
> My proposal is to move to merge this PR, then that that come from https://github.com/luipir/trends.earth/tree/dataset_delete (#416) and at the end that that manage dataset_add where I added the fix you asked

@luipir yes we should move forward with this as I also need to use the changes in those two other PRs in order to complete PR for https://github.com/ConservationInternational/trends.earth/issues/407

**Comment by luipir** [comment]
> @luipir yes we should move forward with this as I also need to use the changes in those two other PRs in order to complete PR for #407

I agree... wait for review or thumb up by @ricardogsilva. If green light, please @Samweli provide to merge if there are no other modifications

**Review by ricardogsilva — APPROVED** [review]
Seems to be working OK now, so I'm +1 for merging.

Just take care of fixing the conflicts reported by github, as it seems to not allow auto merge with the current state

**Comment by ricardogsilva** [comment]
@Samweli I see this still needs you review and approval?

**Comment by Samweli** [comment]
@ricardogsilva I have already reviewed it. @luipir can you update the PR with the latest changes from the decision-trees branch. Then we can merge it.

**Comment by luipir** [comment]
> @ricardogsilva I have already reviewed it. @luipir can you update the PR with the latest changes from the decision-trees branch. Then we can merge it.

doing now

**Comment by ricardogsilva** [comment]
@Samweli if you already reviewed, then please use the github review feature and mark the PR as accepted

**Comment by luipir** [comment]
@Samweli rebased over latest decision-tree updates

**Comment by luipir** [comment]
@samweli after accept review, please merge

**Review by Samweli — APPROVED** [review]

**Pull request merged**
